File: lib/src/models/kalimah.dart
import 'package:sqflite/sqflite.dart';

class Kalimah {
  int id;
  int ayahId;
  String ayahKey;
  int position;
  String? audioUrl;
  String location;
  String textUthmani;
  String codeV1;
  String codeV2;
  String qpcUthmaniHafs;
  String charTypeName;
  int pageNumberV1;
  int pageNumberV2;
  int lineNumberV1;
  int lineNumberV2;
  String translation;
  String? transliteration;

  Kalimah({
    required this.id,
    required this.ayahId,
    required this.ayahKey,
    required this.position,
    this.audioUrl,
    required this.location,
    required this.textUthmani,
    required this.codeV1,
    required this.codeV2,
    required this.qpcUthmaniHafs,
    required this.charTypeName,
    required this.pageNumberV1,
    required this.pageNumberV2,
    required this.lineNumberV1,
    required this.lineNumberV2,
    required this.translation,
    this.transliteration,
  });

  // Konversi objek Kalimah menjadi Map untuk penyimpanan ke database
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'ayah_id': ayahId,
      'ayah_key': ayahKey,
      'position': position,
      'audio_url': audioUrl,
      'location': location,
      'text_uthmani': textUthmani,
      'code_v1': codeV1,
      'code_v2': codeV2,
      'qpc_uthmani_hafs': qpcUthmaniHafs,
      'char_type_name': charTypeName,
      'page_number_v1': pageNumberV1,
      'page_number_v2': pageNumberV2,
      'line_number_v1': lineNumberV1,
      'line_number_v2': lineNumberV2,
      'translation': translation,
      'transliteration': transliteration,
    };
  }

  // Konversi Map menjadi objek Kalimah
  factory Kalimah.fromMap(Map<String, dynamic> map) {
    return Kalimah(
      id: map['id'],
      ayahId: map['ayah_id'],
      ayahKey: map['ayah_key'],
      position: map['position'],
      audioUrl: map['audio_url'],
      location: map['location'],
      textUthmani: map['text_uthmani'],
      codeV1: map['code_v1'],
      codeV2: map['code_v2'],
      qpcUthmaniHafs: map['qpc_uthmani_hafs'],
      charTypeName: map['char_type_name'],
      pageNumberV1: map['page_number_v1'],
      pageNumberV2: map['page_number_v2'],
      lineNumberV1: map['line_number_v1'],
      lineNumberV2: map['line_number_v2'],
      translation: map['translation'],
      transliteration: map['transliteration'],
    );
  }
  // Tambahan operasi CRUD lainnya dapat ditambahkan sesuai kebutuhan
}

File: assets/databases/quran-fetcher/kalimah.py
import requests
import sqlite3
import time

# Function to create SQLite tables
def create_tables():
    conn = sqlite3.connect('quran.db')
    c = conn.cursor()
    c.execute('DROP TABLE IF EXISTS Kalimah')
    c.execute('''CREATE TABLE IF NOT EXISTS Kalimah (
                    id INTEGER PRIMARY KEY,
                    ayah_id INTEGER NOT NULL,
                    ayah_key TEXT NOT NULL,
                    position INTEGER NOT NULL,
                    audio_url TEXT,
                    location TEXT NOT NULL,
                    text_uthmani TEXT NOT NULL,
                    code_v1 TEXT NOT NULL,
                    code_v2 TEXT NOT NULL,
                    qpc_uthmani_hafs TEXT NOT NULL,
                    char_type_name TEXT NOT NULL,
                    page_number_v1 INTEGER NOT NULL,
                    page_number_v2 INTEGER NOT NULL,
                    line_number_v1 INTEGER NOT NULL,
                    line_number_v2 INTEGER NOT NULL,
                    translation TEXT NOT NULL,
                    transliteration TEXT,
                    FOREIGN KEY (ayah_id) REFERENCES Ayah(id)
                )''')
    conn.commit()
    conn.close()
    print("Table Kalimah created")

# Function to insert data into SQLite table
def insert_data(words, verse_id, verse_key):
    print(f"Trying to insert {verse_id} data")
    conn = sqlite3.connect('quran.db')
    c = conn.cursor()
    for word in words:
        c.execute('''INSERT INTO Kalimah (ayah_id, position, audio_url, ayah_key, location, text_uthmani, code_v1, code_v2, qpc_uthmani_hafs, char_type_name, page_number_v1, page_number_v2, line_number_v1, line_number_v2, translation, transliteration)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)''', 
                    (verse_id, word['position'], word['audio_url'], verse_key, word['location'], word['text_uthmani'], 
                    word['code_v1'], word['code_v2'], word['qpc_uthmani_hafs'], word['char_type_name'], word['page_number'], word['page_number'],
                    word['line_number'], word['line_number'], word['translation']['text'], word['transliteration']['text']))
    conn.commit()
    conn.close()
    print(f"Data {verse_id} inserted")

# Main function
def main():
    create_tables()

    for i in range(1, 115):  # Loop through each surah (1 to 114)
        url = f'https://api.qurancdn.com/api/qdc/verses/by_chapter/{i}?words=true&translation_fields=resource_name%2Clanguage_id&per_page=300&fields=text_uthmani%2Cchapter_id%2Chizb_number%2Ctext_imlaei_simple&translations=134&reciter=7&word_translation_language=id&page=1&word_fields=verse_key%2Cverse_id%2Cpage_number%2Clocation%2Ctext_uthmani%2Ccode_v2%2Ccode_v1%2Cqpc_uthmani_hafs%2Ctranslation%2Ctransliteration&mushaf=2'
        data = fetch_data(url)
        print(f"Data {i} fetched")
        for verse in data['verses']:
            insert_data(verse['words'], verse['id'], verse['verse_key'])
        
        time.sleep(0.5)  # Delay 0.5 seconds before fetching data for the next surah

    print("Data inserted successfully.")

# Fetch data from API
def fetch_data(url):
    response = requests.get(url)
    data = response.json()
    return data

if __name__ == "__main__":
    main()
